### ISSUE TITLE
Memory event example freezes the guest

### ISSUE THREAD

**Issue by Luluno01** [libvmi/libvmi#1000]
## Environment

* Intel(R) Xeon(R) CPU E5-2620 v4
* Xen 4.11.4-pre (Ubuntu 4.11.3+24-g14b62ab3e5-1ubuntu2) (ubuntu-devel-discuss@lists.ubuntu.com) (gcc (Ubuntu 9.2.1-31ubuntu3) 9.2.1 20200306) debug=n  Tue Mar 10 09:04:06 UTC 2020
* libvmi commit 17190c623941f3b84baaf1e6ec00d902ff207e15
* Guest kernel (1): Linux v5.15 built from source
* Guest kernel (2): v5.10.0-9 from debian-live-11.1.0-amd64-gnome+nonfree.iso

## Reproducing Steps

1. Build libvmi from source and install
2. Boot guest OS
3. Keep the guest idle
4. Run `mem-event-example <guest name>`

## Expectation

* The guest runs normally (presumably slower but that's fine)
* The example program prints out **different** executed instruction addresses which imply a **normal** execution trace

## Actual Result

* The guest froze immediately, not responding to network traffic or mouse/keyboard input, not updating the screen
* The example program successfully registered the memory event and was receiving events, but all the events were tied to the same instruction address
* `xl list` showed the guest was in an unknown state (`------`)

## Extra Information

For the v5.15 kernel, I `addr2line`-ed the address and it is `kernel/locking/spinlock.c:153`. Does this mean the memory event is somehow breaking the spinlock?

Any ideas?

**Comment by Luluno01** [comment]
I don't really know the internal of how libvmi emits events. All I know is that libvmi removes the corresponding access mode from the guest's physical memory frame to trigger an access violation event when the guest is trying to access that memory region with the specified access mode. So it is possible that `mem-event-example` forgets to restart/emulate the access?

**Comment by hinoue** [comment]
I see the same behavior and am interested in a solution, also.  Should the callback return `VMI_EVENT_RESPONSE_TOGGLE_SINGLESTEP` instead of `VMI_EVENT_RESPONSE_NONE`?

**Comment by Luluno01** [comment]
> I see the same behavior and am interested in a solution, also. Should the callback return `VMI_EVENT_RESPONSE_TOGGLE_SINGLESTEP` instead of `VMI_EVENT_RESPONSE_NONE`?

Yeah, that's indeed a suspicious return value because it sounds like telling Xen to "do not restart the faulted instruction and also do nothing else" - so `VMI_EVENT_RESPONSE_NONE` effectively makes the faulted instruction a `NOP` (or never restarted) I suppose? I haven't tried other return values yet. Maybe I'll do that later :P.

**Comment by hinoue** [comment]
I changed the response to `TOGGLE_SINGLESTEP` and added a single step event and callback.  It was never called.  However, I did get a second address in the output.  It alternated between two addresses (still in the spin-lock?  I do not have debug symbols on that system yet).  The guest still appeared stuck, though.  I'm not sure what to try next.

**Comment by Luluno01** [comment]
I did occasionally get multiple addresses (yes, still in the spinlock in my case) but I am not sure how to reproduce it. Maybe just interrupt the example program and rerun it immediately? I think we would need to figure out if the faulted instruction needs to be explicitly restarted, and if it does, what response we should send back to Xen. Perhaps we need to ask Xen developers, or turn to the libxenctrl doc, and review the event handling code of LibVMI. (That's quite a lot of work to do. Sigh*)



---- Replied mail ----
| From | Hajime (Jim) ***@***.***> |
| Date | 12/01/2021 01:57 |
| To | ***@***.***> |
| Cc | ***@***.******@***.***> |
| Subject | Re: [libvmi/libvmi] Memory event example freezes the guest (Issue #1000) |

I changed the response to TOGGLE_SINGLESTEP and added a single step event and callback. It was never called. However, I did get a second address in the output. It alternated between two addresses (still in the spin-lock? I do not have debug symbols on that system yet). The guest still appeared stuck, though. I'm not sure what to try next.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by Luluno01** [comment]
Found [this](https://xenproject.org/2016/04/13/stealthy-monitoring-with-xen-altp2m/). Looks like if `mem-event-example` is using altp2m ("stealthy monitoring", read-only & non-intrusive property implied I suppose?), we don't need to care about the emulation. But yeah, still not sure what's happening under the hood.

**Comment by hinoue** [comment]
Well, it turns out that I noticed this in the spring.  I had to drop it for several months, and then forgot I looked at it.
Tamas Lengyel pointed out the correct solution then:

> The instruction pointer can't advance while the page access is
restricted. You need to lift the page permission so the CPU can fetch
the instruction and continue. If you don't, it will constantly fault
as you observed. You would normally enable singlestepping in the mem
event callback, lift page permissions, then in the singlestep callback
put the mem event back.

I think the example should be expanded to include this, because it's currently misleading.

**Comment by Luluno01** [comment]
Yes this is exactly what's said [here](https://xenproject.org/2016/04/13/stealthy-monitoring-with-xen-altp2m/) as

> ..., the memory permission needs to be relaxed in order to allow the guest to continue execution.

And this is true for the conventional introspection approach. However, if altp2m is used, manual single-stepping should be unnecessary. So I added debug output to [this if-statement](https://github.com/libvmi/libvmi/blob/master/libvmi/driver/xen/xen_events.c#L59) and enabled `VMI_DEBUG_XEN` output. Interestingly, it turns out `mem-event-example` is indeed using the legacy `set_mem_access` method that requires the user to explicitly "relax the memory permission" and advance the execution, which is not done in this example program. Another finding is that `event-example` is also using legacy `set_mem_access` but it doesn't hang the guest domain. I can see from the debug message, `event-example` is correctly single-stepping the faulted CPU. I'm still looking into these two examples

**Comment by Luluno01** [comment]
Figured out. `mem-event-example` is missing 2 LOC in the callback to allow the faulted CPU to advance. I found these 2 vital lines of code from `event-example`.

Now I am wondering how to use the altp2m version to monitor the memory access more efficiently. It seems we have to specify a `slat_id` for the registered event but I have no idea what it is LOL.

**Comment by hinoue** [comment]
Very nice.  I agree that it would be good to find a more efficient solution.  I notice that guest does in fact advance,
but it does run many thousands of times slower.

**Comment by Luluno01** [comment]
Okay SLAT is "Second Level Address Translation" according to [`slat.h`](https://github.com/libvmi/libvmi/blob/master/libvmi/slat.h#L27). If SLAT is EPT mentioned [here](https://xenproject.org/2016/04/13/stealthy-monitoring-with-xen-altp2m/), then it seems we have to manually create and maintain multiple SLATs - sounds like a lot more work to do :(.

**Comment by julie-nga** [comment]
Hello everyone, 
I’m encountering a similar issue as described here, where my guest VM pauses while running the `mem-event-example` on my VM. 
I noticed that this issue was reported three years ago, and since then, there have been updates to the `mem-event-example` to solve it. However, I’m still experiencing this behavior with the latest version of libvmi.

What could be causing the VM to pause unexpectedly? 
Has anyone successfully run `mem-event-example` on a similar configuration recently?

_**Setup:**_
**Host OS:** Windows 11
**VM:** Debian 11 (via Vagrant)
**LibVMI version:** Latest from GitHub
**Guest VM:** Windows XP
**Vagrant Setup:** Followed instructions from [https://kvm-vmi.github.io/kvm-vmi/master/setup.html#option-1-vagrant-virtual-machine-based-setup](https://kvm-vmi.github.io/kvm-vmi/master/setup.html#option-1-vagrant-virtual-machine-based-setup)
